types/block.go

```diff
@@ -118,13 +184,78 @@ func (b *Block) fillHeader() {
 		b.LastCommitHash = b.LastCommit.Hash()
 	}
 	if b.DataHash == nil {
-		b.DataHash = b.Data.Hash()
+		b.fillDataAvailabilityHeader()
 	}
 	if b.EvidenceHash == nil {
 		b.EvidenceHash = b.Evidence.Hash()
 	}
 }
 
+// fillDataAvailabilityHeader fills in any remaining DataAvailabilityHeader fields
+// that are a function of the block data.
+func (b *Block) fillDataAvailabilityHeader() {
+	namespacedShares := b.Data.computeShares()
+	shares := namespacedShares.RawShares()
+	if len(shares) == 0 {
+		// no shares -> no row/colum roots -> hash(empty)
+		b.DataHash = b.DataAvailabilityHeader.Hash()
+		return
+	}
+	// TODO(ismail): for better efficiency and a larger number shares
+	// we should switch to the rsmt2d.LeopardFF16 codec:
+	extendedDataSquare, err := rsmt2d.ComputeExtendedDataSquare(shares, rsmt2d.RSGF8)
+	if err != nil {
+		panic(fmt.Sprintf("unexpected error: %v", err))
+	}
+	// compute roots:
+	squareWidth := extendedDataSquare.Width()
+	originalDataWidth := squareWidth / 2
+	b.DataAvailabilityHeader = DataAvailabilityHeader{
+		RowsRoots:   make([]namespace.IntervalDigest, squareWidth),
+		ColumnRoots: make([]namespace.IntervalDigest, squareWidth),
+	}
+
+	// compute row and column roots:
```

Comment:
This will need to be refactored as https://github.com/lazyledger/rsmt2d/blob/master/extendeddatacrossword.go uses RowRoots() and ColumnRoots() in the data square itself to reconstruct the square.